File: android/src/main/java/com/mapbox/mapboxgl/AnimatedRoute.java
package com.mapbox.mapboxgl;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.graphics.Color;

import com.mapbox.geojson.Feature;
import com.mapbox.geojson.LineString;
import com.mapbox.geojson.Point;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.mapboxsdk.style.layers.LineLayer;
import com.mapbox.mapboxsdk.style.layers.Property;
import com.mapbox.mapboxsdk.style.sources.GeoJsonSource;
import com.mapbox.turf.TurfMeasurement;
import com.mapbox.turf.TurfMisc;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineCap;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineColor;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineJoin;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineWidth;

public class AnimatedRoute {

    private static final String LINE_SOURCE_ID = "line-source-id";
    private static final String TARGET_LINE_SOURCE_ID = "target-line-source-id";
    private static final String TARGET_LINE_LAYER_ID = "target-line-layer-id";

    private MapboxMap mapInstance;
    private Style style;
    private AnimatedMarker animatedMarker;

    private GeoJsonSource lineSource;
    private LineString route;
    private List<Point> routeCoordinateList;

    private GeoJsonSource targetLineSource;
    private LineString targetRoute;
    private List<Point> targetRouteCoordinateList;

    private float bearingPrev = 0;
    boolean isRotating = false;

    private int routeIndex;

    public AnimatedRoute(MapboxMap mapView, AnimatedMarker marker) {
        mapInstance = mapView;
        style = Objects.requireNonNull(mapInstance.getStyle());
        animatedMarker = marker;

        createSources();
        createTargetLine();
    }

    private void createSources() {
        style.addSource(lineSource = new GeoJsonSource(LINE_SOURCE_ID));
        style.addSource(targetLineSource = new GeoJsonSource(TARGET_LINE_SOURCE_ID));
    }

    private void createTargetLine() {
        style.addLayerBelow(new LineLayer(TARGET_LINE_LAYER_ID, TARGET_LINE_SOURCE_ID).withProperties(
                lineColor(Color.parseColor("#000000")),
                lineCap(Property.LINE_CAP_ROUND),
                lineJoin(Property.LINE_JOIN_ROUND),
                lineWidth(2.5f)), "road-label");
    }

    public void update(AnimatedRouteConfiguration conf) {
        route = LineString.fromPolyline(conf.line, 6);
        targetRoute = LineString.fromPolyline(conf.targetLine, 6);

        routeCoordinateList = route.coordinates();
        final LineString slicedSource = animatedMarker.currentLocation().equals(routeCoordinateList.get(routeCoordinateList.size() - 1)) ?
                route :
                TurfMisc.lineSlice(animatedMarker.currentLocation(), routeCoordinateList.get(routeCoordinateList.size() - 1), route);

        route = slicedSource;
        routeCoordinateList = slicedSource.coordinates();

        if (!animatedMarker.currentLocation().equals(routeCoordinateList.get(0))) {
            routeCoordinateList.add(0, animatedMarker.currentLocation());
        }

        targetRouteCoordinateList = targetRoute.coordinates();
        routeIndex = 0;

        animate();

        lineSource.setGeoJson(route);
        targetLineSource.setGeoJson(targetRoute);
    }

    private void animate() {
        if (routeCoordinateList.get(0).equals(routeCoordinateList.get(routeCoordinateList.size() - 1))) {
            final List<Point> list = new ArrayList<Point>();
            list.add(routeCoordinateList.get(0));
            list.add(routeCoordinateList.get(routeCoordinateList.size() - 1));

            routeCoordinateList = list;
        }

        // Check if we are at the end of the points list
        if ((routeCoordinateList.size() - 1 > routeIndex)) {
            final Point currentPosition = animatedMarker.currentLocation();
            final Point targetPosition = routeCoordinateList.get(routeIndex + 1);

            final long size = (long) TurfMeasurement.distance(routeCoordinateList.get(0), routeCoordinateList.get(routeCoordinateList.size() - 1), "meters");

            if (size > 1000) {
                animatedMarker.updateCoordinates(routeCoordinateList.get(routeCoordinateList.size() - 1), 300);
                return;
            }

            /*targetLineSource.setGeoJson(
                    Feature.fromGeometry(
                            TurfMisc.lineSlice(targetPosition, targetRouteCoordinateList.get(targetRouteCoordinateList.size() - 1), targetRoute)
                    )
            );*/

            long speed;

            if (size > 300) {
                speed = 10;
            } else if (size > 150) {
                speed = 20;
            } else if (size > 100) {
                speed = 35;
            } else if (size > 55) {
                speed = 105;
            } else if (size > 30) {
                speed = 135;
            } else {
                speed = 160;
            }

            long meters = (long) TurfMeasurement.distance(routeCoordinateList.get(routeIndex), targetPosition, "meters");
            long duration = speed * meters;

            // float bearing = (float) TurfMeasurement.bearing(currentPosition, targetPosition) % 360;
            // float bearing = (float) TurfMeasurement.bearing(currentPosition, targetPosition);
            /*if (bearing < 0) {
                bearing += 360;
            }*/

            float bearing = (float) getBearing(currentPosition, targetPosition);

            animatedMarker.updateCoordinates(targetPosition, duration, new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);

                    animate();
                }
            });

            animatedMarker.updateRotation(bearing, 200);

            routeIndex++;
        }
    }

    private double getBearing(Point currentPosition, Point targetPosition) {
        float bearing = (float) TurfMeasurement.bearing(currentPosition, targetPosition) % 360;
        if (bearing < 0) {
            bearing += 360;
        }

        float currentValue = animatedMarker.prevBearing;

        float nextBearing = bearing;

        if (Math.abs(bearing - animatedMarker.prevBearing) > 150) {
            if (animatedMarker.prevBearing >= 180 && bearing < 180) {
                currentValue = -(360 - animatedMarker.prevBearing);
            }

            if (animatedMarker.prevBearing <= 180 && bearing > 180) {
                nextBearing = -(360 - bearing);
            }
        }

        final float pBearing = currentValue;
        final float nBearing = nextBearing;

        animatedMarker.prevBearing = pBearing;

        return nBearing;
    }

    public void destroy() {
        style.removeLayer(TARGET_LINE_LAYER_ID);
        style.removeSource(LINE_SOURCE_ID);
        style.removeSource(TARGET_LINE_SOURCE_ID);
    }

}

File: ios/Classes/AnimatedRoute.swift
//
//  AnimatedRoute.swift
//  mapbox_gl
//
//  Created by Eugene Pisotsky on 12.08.2020.
//

import Foundation
import Mapbox
import Turf

class AnimatedRouteConfiguration {
    var line: String
    var targetLine: String
    
    init(line: String, targetLine: String) {
        self.line = line
        self.targetLine = targetLine
    }
}

class AnimatedRoute {
    
    var LINE_SOURCE_ID = "line-source-id"
    var TARGET_LINE_SOURCE_ID = "target-line-source-id"
    var TARGET_LINE_LAYER_ID = "target-line-layer-id"
    
    var routeIndex: Int = 0
    
    var route: MGLPolyline?
    var targetRoute: MGLPolyline?
    var lineSource: MGLShapeSource?
    var targetLineSource: MGLShapeSource?
    var routeCoordinates: [CLLocationCoordinate2D]?
    var targetRouteCoordinates: [CLLocationCoordinate2D]?
    var targetLineLayer: MGLLineStyleLayer?
    
    var mapInstance: MGLMapView
    var animatedMarker: AnimatedMarker
    
    init(mapInstance: MGLMapView, marker: AnimatedMarker) {
        self.mapInstance = mapInstance
        self.animatedMarker = marker
        
        createSources()
        createTargetLine()
    }
    
    func createSources() {
        targetLineSource = MGLShapeSource(identifier: TARGET_LINE_SOURCE_ID, shape: targetRoute)
        lineSource = MGLShapeSource(identifier: LINE_SOURCE_ID, shape: route)
        mapInstance.style?.addSource(targetLineSource!)
        mapInstance.style?.addSource(lineSource!)
    }

    func createTargetLine() {
        let layer = MGLLineStyleLayer(identifier: TARGET_LINE_LAYER_ID, source: targetLineSource!)
        
        targetLineLayer = layer
        
        layer.lineColor = NSExpression(forConstantValue: UIColor.black)
        layer.lineCap = NSExpression(forConstantValue: "round")
        layer.lineJoin = NSExpression(forConstantValue: "round")
        layer.lineWidth = NSExpression(forConstantValue: 2.5)
        
        mapInstance.style?.addLayer(layer)
    }
    
    func update(conf: AnimatedRouteConfiguration) {
        let polyline = Polyline(encodedPolyline: conf.line, precision: 1e6)
        guard let decodedCoordinates: [CLLocationCoordinate2D] = polyline.coordinates else { return }
        route = MGLPolyline(coordinates: decodedCoordinates, count: UInt(decodedCoordinates.count))
        
        let targetPolyline = Polyline(encodedPolyline: conf.targetLine, precision: 1e6)
        guard let targetDecodedCoordinates: [CLLocationCoordinate2D] = targetPolyline.coordinates else { return }
        
        routeCoordinates = decodedCoordinates
        targetRouteCoordinates = targetDecodedCoordinates
        
        let currentCoord = animatedMarker.currentLocation()
        animatedMarker.annotation?.coordinate = currentCoord!
        
        if (currentCoord != routeCoordinates?.last) {
            let slicedRoute = LineString(routeCoordinates!).sliced(from: currentCoord, to: routeCoordinates?.last)
            
            let coords: [CLLocationCoordinate2D] = slicedRoute?.coordinates ?? []
            
            route = MGLPolyline(coordinates: coords, count: UInt(coords.count))
            routeCoordinates = coords
        }
        
        routeIndex = 0
        animate()
        
        lineSource?.shape = MGLPolylineFeature(coordinates: routeCoordinates!, count: UInt(routeCoordinates!.count))
        targetLineSource?.shape = MGLPolylineFeature(coordinates: targetRouteCoordinates!, count: UInt(targetRouteCoordinates!.count))
    }
    
    func animate() {
        if (routeCoordinates?.first == routeCoordinates?.last) {
            let newList: [CLLocationCoordinate2D]? = [
                (routeCoordinates?.first)!,
                (routeCoordinates?.last)!
            ]

            routeCoordinates = newList
        }
        
        guard let list = self.routeCoordinates else { return }
        
        if (list.count <= self.routeIndex + 1) {
            return
        }

        let size = list.first!.distance(to: list.last!)
        let sizeForStep = list[self.routeIndex].distance(to: list[self.routeIndex + 1])
        var speed: Double
        
        if (size > 1000) {
            guard let lastCoord = routeCoordinates?.last else { return }
            animatedMarker.updateCoordinates(coords: lastCoord, duration: 0.3)
            
            return
        }
        
        if (size > 300) {
            speed = 0.01
        } else if (size > 150) {
            speed = 0.02
        } else if (size > 100) {
            speed = 0.035
        } else if (size > 55) {
            speed = 0.105
        } else if (size > 30) {
            speed = 0.135
        } else {
            speed = 0.160
        }
        
        speed = speed * sizeForStep
        
        if ((routeCoordinates?.count ?? 0) - 1 > routeIndex) {
            let prevCoords = list[self.routeIndex]
            let coords = list[self.routeIndex + 1]
            
            if (prevCoords != coords) {
                let bearing = getBearingBetweenTwoPoints1(point1: prevCoords, point2: coords)
                animatedMarker.updateRotation(rotation: bearing, duration: 2.0)
            }
            
            animatedMarker.updateCoordinates(coords: coords, duration: speed)
            self.routeIndex += 1
            
            animatedMarker.animationCoord?.addCompletion { position in
                if position == .end {
                    self.animate()
                }
            }
        }
    }
    
    func getBearingBetweenTwoPoints1(point1 : CLLocationCoordinate2D, point2 : CLLocationCoordinate2D) -> Double {
        let lat1 = degreesToRadians(degrees: point1.latitude)
        let lon1 = degreesToRadians(degrees: point1.longitude)

        let lat2 = degreesToRadians(degrees: point2.latitude)
        let lon2 = degreesToRadians(degrees: point2.longitude)

        let dLon = lon2 - lon1

        let y = sin(dLon) * cos(lat2)
        let x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(dLon)
        let radiansBearing = atan2(y, x)

        return radiansToDegrees(radians: radiansBearing)
    }

    
    func degreesToRadians(degrees: Double) -> Double { return degrees * .pi / 180.0 }
    func radiansToDegrees(radians: Double) -> Double { return radians * 180.0 / .pi }
    
    func destroy() {
        mapInstance.style?.removeLayer(targetLineLayer!)
    }
    
}
